### ISSUE TITLE
Julia 1.6 RC1 segmentation fault when running JLD.save()

### ISSUE THREAD

**Issue by StatisticalMice** [JuliaLang/julia#39655]
The Dockerfile for the environment where the segmentation fault  happens:
```
FROM julia:1.6-buster

USER root
RUN useradd -ms /bin/bash arthur

USER arthur
WORKDIR /home/arthur
ENV PATH="/home/arthur/mambaforge/bin:${PATH}"

RUN set -eux; \
    curl -LJO --output Mambaforge-Linux-x86_64.sh \
    https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh \
    && bash Mambaforge-Linux-x86_64.sh -b \
    && rm -f Mambaforge-Linux-x86_64.sh 

RUN set -eux; \
    mamba install -y jupyterlab \
    && mamba clean --all -f -y

RUN julia -e "using Pkg; Pkg.add([\"JLD\",\"SnoopCompile\",\"AbstractTrees\"]); Pkg.precompile();"

CMD ["bash"]
```

Code run
```
using JLD, SnoopCompile, Pkg
tinf = @snoopi_deep Pkg.add("DataFrames")
JLD.save("./shared/tinf.jld", "tinf", tinf)
```

Output
```
julia> tinf
InferenceTimingNode: 37.478120/38.527522 on InferenceFrameInfo for Core.Compiler.Timings.ROOT() with 78 direct children

julia> JLD.save("./shared/tinf.jld", "tinf", tinf)

signal (11): Segmentation fault
in expression starting at REPL[8]:1
typekeyvalue_hash at /buildworker/worker/package_linux64/build/src/jltypes.c:1006 [inlined]
lookup_typevalue at /buildworker/worker/package_linux64/build/src/jltypes.c:600
lookup_arg_type_tuple at /buildworker/worker/package_linux64/build/src/gf.c:1842 [inlined]
jl_lookup_generic_ at /buildworker/worker/package_linux64/build/src/gf.c:2354 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2416
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_apply at /buildworker/worker/package_linux64/build/src/builtins.c:672
writeas at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:869 [inlined]
write_ref at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:691
macro expansion at /home/arthur/.julia/packages/JLD/nQ9iW/src/jld_types.jl:639 [inlined]
h5convert! at /home/arthur/.julia/packages/JLD/nQ9iW/src/jld_types.jl:654
unknown function (ip: 0x7fb256efae54)
#write_compound#24 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:733
write_compound at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:727 [inlined]
#_write#23 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:723 [inlined]
_write at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:723 [inlined]
write_ref at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:691
macro expansion at /home/arthur/.julia/packages/JLD/nQ9iW/src/jld_types.jl:639 [inlined]
h5convert! at /home/arthur/.julia/packages/JLD/nQ9iW/src/jld_types.jl:654
unknown function (ip: 0x7fb256ef74c4)
#write_compound#24 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:733
write_compound at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:727 [inlined]
#_write#23 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:723 [inlined]
_write at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:723 [inlined]
write_ref at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:691
macro expansion at /home/arthur/.julia/packages/JLD/nQ9iW/src/jld_types.jl:639 [inlined]
h5convert! at /home/arthur/.julia/packages/JLD/nQ9iW/src/jld_types.jl:654
unknown function (ip: 0x7fb256ef33db)
#write_compound#24 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:733
write_compound at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:727 [inlined]
#_write#23 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:723 [inlined]
_write at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:723 [inlined]
write_ref at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:691
macro expansion at /home/arthur/.julia/packages/JLD/nQ9iW/src/jld_types.jl:639 [inlined]
h5convert! at /home/arthur/.julia/packages/JLD/nQ9iW/src/jld_types.jl:654
unknown function (ip: 0x7fb256ef00cb)
#write_compound#24 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:733
write_compound at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:727 [inlined]
#_write#23 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:723 [inlined]
_write at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:723 [inlined]
write_ref at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:691
macro expansion at /home/arthur/.julia/packages/JLD/nQ9iW/src/jld_types.jl:639 [inlined]
h5convert! at /home/arthur/.julia/packages/JLD/nQ9iW/src/jld_types.jl:654
unknown function (ip: 0x7fb256eeca9b)
#write_compound#24 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:733
write_compound at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:727 [inlined]
#_write#23 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:723 [inlined]
_write at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:723 [inlined]
#write#17 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:542 [inlined]
write at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:542 [inlined]
#39 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:1260
#jldopen#14 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:261
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
jldopen##kw at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:259
unknown function (ip: 0x7fb256eb9881)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
#save#38 at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:1258
save at /home/arthur/.julia/packages/JLD/nQ9iW/src/JLD.jl:1255
unknown function (ip: 0x7fb256eb96e3)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_apply at /buildworker/worker/package_linux64/build/src/builtins.c:672
#save#19 at /home/arthur/.julia/packages/FileIO/TyKdX/src/loadsave.jl:139
save at /home/arthur/.julia/packages/FileIO/TyKdX/src/loadsave.jl:139
unknown function (ip: 0x7fb256eb92ac)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:115
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:204
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:155 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:561
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:669
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:879
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:827
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:931
eval at ./boot.jl:360
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
eval_user_input at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:139
repl_backend_loop at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:200
start_repl_backend at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:185
#run_repl#42 at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:317
run_repl at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:305
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
#878 at ./client.jl:387
jfptr_YY.878_47637.clone_1 at /usr/local/julia/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
do_apply at /buildworker/worker/package_linux64/build/src/builtins.c:672
jl_f__apply_latest at /buildworker/worker/package_linux64/build/src/builtins.c:722
#invokelatest#2 at ./essentials.jl:707 [inlined]
invokelatest at ./essentials.jl:706 [inlined]
run_main_repl at ./client.jl:372
exec_options at ./client.jl:302
_start at ./client.jl:485
jfptr__start_50107.clone_1 at /usr/local/julia/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2238 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2420
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1703 [inlined]
true_main at /buildworker/worker/package_linux64/build/src/jlapi.c:557
repl_entrypoint at /buildworker/worker/package_linux64/build/src/jlapi.c:699
main at julia (unknown line)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x4007d8)
Allocations: 23555771 (Pool: 23545789; Big: 9982); GC: 30
Segmentation fault
arthur@db7f566454d7:~$ 
```

Invoked with 
```
docker build -t tmp .
docker run -ti --rm  -v /Volumes/MyFiles/GitRepos/StatisticalMice/tmp:/home/arthur/shared tmp
```

```
arthur@c2314865bea0:~$ julia --version
julia version 1.6.0-rc1
```

**Comment by StatisticalMice** [comment]
This is a simpler version to reproduce it with less moving parts.
```
FROM julia:1.6-buster

USER root
RUN useradd -ms /bin/bash arthur

USER arthur
WORKDIR /home/arthur

RUN julia -e "using Pkg; Pkg.add([\"JLD\",\"SnoopCompile\",\"AbstractTrees\",\"Profile\",\"PProf\"]); Pkg.precompile();"

CMD ["bash"]
```

```
using JLD, SnoopCompile, Pkg
tinf = @snoopi_deep Pkg.add("DataFrames")
JLD.save("tinf.jld", "tinf", tinf)
```

**Comment by JeffBezanson** [comment]
It looks like this comes from trying to save a SimpleVector with uninitialized entries, which exist in the `specializations` field of `Method`s sometimes.

**Comment by JeffBezanson** [comment]
Is this known to have worked on 1.5? I'm not sure it's a regression.

**Comment by JeffBezanson** [comment]
Next issue is that I think there is a reference cycle that JLD doesn't handle.

**Comment by StatisticalMice** [comment]
@JeffBezanson The reason I tried doing this is that I tried to visualise the results as described in [Profiling type-inference](https://julialang.org/blog/2021/01/snoopi_deep/), but the code ran in a container without a GUI. It made sense to export the results to the host machine that would be able to visualise them.

**Comment by vtjnash** [comment]
Closed by #39695. I can't promise it works, since Jeff mentioned JLD doesn't support this, but the error in Base causing a segfault should at least be addressed.
